### PR TITLE
Add support for client name configuration

### PR DESCRIPTION
**Pull request by ikolomi**
*Issue #, if available:*
Part of https://github.com/aws/babushka/issues/407

*Description of changes:*
Add optional client_name property to RedisConnectionInfo, which will be used with 'CLIENT SETNAME' command during connection setup.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: redis/src/aio/mod.rs [hunk 1]
```diff
@@ -140,6 +140,20 @@ where
         }
     }
 
+    if connection_info.client_name.is_some() {
+        match cmd("CLIENT")
+            .arg("SETNAME")
+            .arg(connection_info.client_name.as_ref().unwrap())
```
Comment (shachlanAmazon) [inline redis/src/aio/mod.rs:146]: instead of using unwrap, you can use an `if let` expression.

File: redis/src/aio/mod.rs [hunk 2]
```diff
@@ -140,6 +140,20 @@ where
...
+            .query_async(con).await
+        {
+            Ok(Value::Okay) => {}
+            _ => fail!((
```
Comment (shachlanAmazon) [inline redis/src/aio/mod.rs:150]: since the returned value of `query_async` is `RedisError`, you don't need to fail, just use the `?` operator.

File: redis/src/aio/mod.rs [hunk 3]
```diff
@@ -140,6 +140,20 @@ where
         }
     }
 
+    if connection_info.client_name.is_some() {
```
Comment (shachlanAmazon) [inline redis/src/aio/mod.rs:143]: this implementation sends the CLIENT SETNAME command and only after it completes the CLIENT SETINFO command is sent. Wouldn't it be better to send both concurrently?

File: redis/src/connection.rs [hunk 4]
```diff
@@ -979,6 +983,20 @@ fn setup_connection(
         }
     }
 
+    if connection_info.client_name.is_some() {
```
Comment (shachlanAmazon) [inline redis/src/connection.rs:986]: same - `if let` and `?`

File: redis/tests/support/mod.rs [hunk 5]
```diff
@@ -349,28 +349,7 @@ impl TestContext {
         Self::with_modules(&[], true)
     }
 
-    pub fn with_tls(tls_files: TlsFilePaths, mtls_enabled: bool) -> TestContext {
```
Comment (shachlanAmazon) [inline redis/tests/support/mod.rs:352]: why did you move this?

File: redis/tests/support/mod.rs [hunk 6]
```diff
@@ -412,30 +416,32 @@ impl TestContext {
...
-                    }
-                }
-                Ok(x) => {
-                    con = x;
-                    break;
-                }
+        TestContext { server, client }
+    }
+
+    pub fn with_clientname(clientname: &str) -> TestContext {
```
Comment (shachlanAmazon) [inline redis/tests/support/mod.rs:424]: client_name

File: redis/tests/test_basic.rs [hunk 7]
```diff
@@ -1448,3 +1449,32 @@ fn test_blocking_sorted_set_api() {
...
+fn test_set_client_name_by_config() {
+    const CLIENT_NAME: &str = "TEST_CLIENT_NAME";
+
+    let ctx = TestContext::with_clientname(CLIENT_NAME);
+    let mut con = ctx.connection();
+
+    let client_list: redis::InfoDict = redis::cmd("CLIENT").arg("LIST").query(&mut con).unwrap();
+    assert!(client_list.len() == 1, "Expecting single client, got: {}", client_list.len());
+
+    let val: &redis::Value = client_list.values().next().unwrap();
```
Comment (shachlanAmazon) [inline redis/tests/test_basic.rs:1463]: can this be parsed to a HashMap, and let the parsing operation find the keys and values?

File: redis/tests/test_basic.rs [hunk 8]
```diff
@@ -1448,3 +1449,32 @@ fn test_blocking_sorted_set_api() {
         );
     }
 }
+
+#[test]
+fn test_set_client_name_by_config() {
+    const CLIENT_NAME: &str = "TEST_CLIENT_NAME";
+
+    let ctx = TestContext::with_clientname(CLIENT_NAME);
+    let mut con = ctx.connection();
```
Comment (shachlanAmazon) [inline redis/tests/test_basic.rs:1458]: notice that you're testing the sync connection, which is the connection we care less about.

**Review by shachlanAmazon — COMMENTED** [review]
round

File: redis/src/aio/mod.rs (hunk 3 above)
Comment (ikolomi) [reply redis/src/aio/mod.rs:143]: You mean add it to client_set_info_pipeline?  yes, it can be done. Why dont we have pipelined DB SELECT then?

File: redis/src/aio/mod.rs (hunk 1 above)
Comment (ikolomi) [reply redis/src/aio/mod.rs:146]: show me please.
Im pretty sure I took this line from other command issuing instance

File: redis/src/aio/mod.rs (hunk 2 above)
Comment (ikolomi) [reply redis/src/aio/mod.rs:150]: Again this is the same as SELECT command just above

File: redis/src/connection.rs (hunk 4 above)
Comment (ikolomi) [reply redis/src/connection.rs:986]: link to an example

File: redis/tests/support/mod.rs (hunk 5 above)
Comment (ikolomi) [reply redis/tests/support/mod.rs:352]: I did not, I moved the common connection establishment code (now in use by 3 funcs) into connect_with_retries()

File: redis/tests/support/mod.rs (hunk 6 above)
Comment (ikolomi) [reply redis/tests/support/mod.rs:428]: will change

File: redis/tests/test_basic.rs (hunk 8 above)
Comment (ikolomi) [reply redis/tests/test_basic.rs:1458]: Yep, I am absolutely aware. The intent is to test the API and the configuration propagation , not the method of invocation. I consider it the same value as the basic get/set tests

File: redis/tests/test_basic.rs (hunk 7 above)
Comment (ikolomi) [reply redis/tests/test_basic.rs:1463]: have no idea what you mean, show me if its better

File: redis/src/aio/mod.rs (hunk 3 above)
Comment (ikolomi) [reply redis/src/aio/mod.rs:143]: will move to client_set_info_pipeline

Comment (ikolomi) [reply redis/src/aio/mod.rs:143]: beyond scope of this PR

File: redis/src/aio/mod.rs (hunk 1 above)
Comment (ikolomi) [reply redis/src/aio/mod.rs:146]: will do if let

File: redis/src/aio/mod.rs (hunk 2 above)
Comment (ikolomi) [reply redis/src/aio/mod.rs:151]: beyond this PR - will refactor in different PR

File: redis/src/connection.rs (hunk 4 above)
Comment (ikolomi) [reply redis/src/connection.rs:986]: will do

File: redis/tests/support/mod.rs (hunk 5 above)
Comment (ikolomi) [reply redis/tests/support/mod.rs:352]: git issue

File: redis/tests/test_basic.rs (hunk 8 above)
Comment (ikolomi) [reply redis/tests/test_basic.rs:1457]: need to add to async also

File: redis/tests/test_basic.rs (hunk 7 above)
Comment (ikolomi) [reply redis/tests/test_basic.rs:1465]: not relevant, will try to look into infoDict

File: redis/src/connection.rs (hunk 4 above)
Comment (ikolomi) [reply redis/src/connection.rs:986]: this non async version of setup connection uses is_some() with the password, will stick to that

File: redis/tests/test_async.rs [hunk 9]
```diff
@@ -781,3 +782,52 @@ mod mtls_test {
...
+        let val: &redis::Value = client_list.values().next().unwrap();
+
+        match val {
+            redis::Value::SimpleString(s) => {
+                let client_attr: Vec<&str> = s.split(' ').collect();
+                for i in client_attr.iter() {
+                    let this_attr: Vec<&str> = i.split('=').collect();
+                    if this_attr[0] == "name" {
+                        assert!(
+                            this_attr[1] == CLIENT_NAME,
```
[the remark this answers is not in the document]
Comment (barshaul) [reply redis/tests/test_async.rs:814]: You can take an example from this test:
https://github.com/amazon-contributing/redis-rs/pull/56/files#diff-85259ca2185589cf8979102fc91c6ce1a9461aac8ed605472b9ad53c3324c400R1912

[the remark this answers is not in the document]
Comment (barshaul) [reply redis/tests/test_async.rs:814]: Summarizing our talk yesterday - the test will be changed to check the client ID and based on the it would check that its name is the one passed in the config

File: redis/tests/test_cluster.rs [hunk 10]
```diff
@@ -887,6 +887,37 @@ fn test_cluster_route_correctly_on_packed_transaction_with_single_node_requests2
...
+#[test]
+fn test_cluster_with_client_name() {
+    let cluster = TestClusterContext::new_with_cluster_client_builder(
+        3,
+        0,
+        |builder| builder.client_name(RedisCluster::client_name().to_string()),
+        false,
+    );
+    let mut con = cluster.connection();
+    let client_info: String = redis::cmd("CLIENT").arg("INFO").query(&mut con).unwrap();
```
Comment (barshaul) [inline redis/tests/test_cluster.rs:899]: i'm not sure if client info is being sent to all nodes, have you checked it?

Comment (barshaul) [reply redis/tests/test_cluster.rs:899]: I think routing it to all nodes would cover the functionality of this feature in cluster better

File: redis/src/cluster_client.rs [hunk 11]
```diff
@@ -33,6 +33,7 @@ struct BuilderParams {
     retries_configuration: RetryParams,
     connection_timeout: Option<Duration>,
     topology_checks_interval: Option<Duration>,
+    client_name: Option<String>,
```
Comment (shachlanAmazon) [inline redis/src/cluster_client.rs:36]: minor: you can use `Option<ArcStr>` throughout, and remove the extra allocations.

File: redis/src/cluster_client.rs [hunk 12]
```diff
@@ -212,6 +222,13 @@ impl ClusterClientBuilder {
                 )));
             }
 
+            if client_name.is_some() && node.redis.client_name != *client_name {
```
Comment (shachlanAmazon) [inline redis/src/cluster_client.rs:225]: IMO we should return an error if `node.redis.client_name` is set, regardless of `client_name` it should be an error, because it might confuse users - they'll think that setting the client name on the initial nodes will affect the whole cluster, but it won't survive any change to the cluster. WDYT?

I see that the same mistake is made with the username and password, so this can be fixed in a separate PR.

File: redis/tests/support/cluster.rs [hunk 13]
```diff
@@ -74,6 +74,10 @@ impl RedisCluster {
         "world"
     }
 
+    pub fn client_name() -> &'static str {
```
Comment (shachlanAmazon) [inline redis/tests/support/cluster.rs:77]: minor - why use a function and not a constant?

File: redis/tests/test_async.rs [hunk 14]
```diff
@@ -781,3 +782,40 @@ mod mtls_test {
...
+            .arg("INFO")
+            .query_async(&mut con)
+            .await
+            .unwrap();
+
+        let client_attrs: Vec<&str> = client_info.split(' ').collect();
+        for attr in client_attrs.iter() {
+            let this_attr: Vec<&str> = attr.split('=').collect();
+            if this_attr[0] == "name" {
+                assert!(
```
Comment (shachlanAmazon) [inline redis/tests/test_async.rs:805]: you can use `assert_eq!`, it logs the different values so there's no need for the extra message.

File: redis/tests/test_basic.rs [hunk 15]
```diff
@@ -1448,3 +1448,32 @@ fn test_blocking_sorted_set_api() {
...
+    let ctx = TestContext::with_client_name(CLIENT_NAME);
+    let mut con = ctx.connection();
+
+    let client_info: String = redis::cmd("CLIENT").arg("INFO").query(&mut con).unwrap();
+
+    let client_attrs: Vec<&str> = client_info.split(' ').collect();
+    for attr in client_attrs.iter() {
+        let this_attr: Vec<&str> = attr.split('=').collect();
+        if this_attr[0] == "name" {
+            assert!(
```
Comment (shachlanAmazon) [inline redis/tests/test_basic.rs:1465]: same

File: redis/tests/test_async.rs [hunk 16]
```diff
@@ -781,3 +782,40 @@ mod mtls_test {
...
+    block_on_all(async move {
+        let mut con = ctx.async_connection().await?;
+
+        let client_info: String = redis::cmd("CLIENT")
+            .arg("INFO")
+            .query_async(&mut con)
+            .await
+            .unwrap();
+
+        let client_attrs: Vec<&str> = client_info.split(' ').collect();
```
Comment (shachlanAmazon) [inline redis/tests/test_async.rs:801]: you're using the same logic, at the very least move the splitting into attribute pairs to a separate function. If you really want to make it handy, have it return `HashMap<String, String>`.

Also, there's no need for the collecting. `for` statements  work on iterators:
```rs
for (key, value) in client_info.split(' ').map(|line| attr.split('=').collect::<(&str, &str)>()) {
```
should be sufficient.

File: redis/tests/test_cluster.rs (hunk 10 above)
Comment (shachlanAmazon) [reply redis/tests/test_cluster.rs:899]: you can't route manually in sync cluster.

Comment (ikolomi) [reply redis/tests/test_cluster.rs:899]: We think that randomly checking a single node is good enough

File: redis/src/cluster_client.rs (hunk 11 above)
Comment (ikolomi) [reply redis/src/cluster_client.rs:36]: Again, I prefer unity overs asking why some of the params (password, username) are by Option and this one is by ArcStr

File: redis/Cargo.toml [hunk 17]
```diff
@@ -137,6 +137,8 @@ tokio = { version = "1", features = ["rt", "macros", "rt-multi-thread", "time"]
 tempfile = "=3.6.0"
 once_cell = "1"
 anyhow = "1"
+sscanf = "0.4.1"
+
```
Comment (shachlanAmazon) [inline redis/Cargo.toml:141]: minor: remove extra line

File: redis/src/cluster_client.rs [hunk 18]
```diff
@@ -191,6 +194,13 @@ impl ClusterClientBuilder {
             };
         }
 
+        let client_name = if cluster_params.client_name.is_none() {
```
Comment (shachlanAmazon) [inline redis/src/cluster_client.rs:197]: remove, as discussed

File: redis/src/cluster_client.rs (hunk 12 above)
Comment (shachlanAmazon) [inline redis/src/cluster_client.rs:225]: I think we said it should be `if client_name != node.redis.client_name {`. rigtht?

File: redis/src/cluster_client.rs [hunk 19]
```diff
@@ -221,6 +238,12 @@ impl ClusterClientBuilder {
         })
     }
 
+    /// Sets password for the new ClusterClient.
```
Comment (shachlanAmazon) [inline redis/src/cluster_client.rs:241]: fix comment

File: redis/tests/test_async.rs [hunk 20]
```diff
@@ -781,3 +782,35 @@ mod mtls_test {
...
+
+        let client_info: String = redis::cmd("CLIENT")
+            .arg("INFO")
+            .query_async(&mut con)
+            .await
+            .unwrap();
+
+        let client_attrs = parse_client_info(&client_info);
+
+        assert!(
```
Comment (shachlanAmazon) [inline redis/tests/test_async.rs:803]: minor: this is essentially contained in the next assert, IMO can be removed

File: redis/tests/test_basic.rs [hunk 21]
```diff
@@ -1448,3 +1448,26 @@ fn test_blocking_sorted_set_api() {
...
+    const CLIENT_NAME: &str = "TEST_CLIENT_NAME";
+
+    let ctx = TestContext::with_client_name(CLIENT_NAME);
+    let mut con = ctx.connection();
+
+    let client_info: String = redis::cmd("CLIENT").arg("INFO").query(&mut con).unwrap();
+
+    let client_attrs = parse_client_info(&client_info);
+
+    assert!(
```
Comment (shachlanAmazon) [inline redis/tests/test_basic.rs:1463]: same

File: redis/tests/test_cluster.rs [hunk 22]
```diff
@@ -887,6 +887,33 @@ fn test_cluster_route_correctly_on_packed_transaction_with_single_node_requests2
...
+
+    let client_attrs = parse_client_info(&client_info);
+
+    assert!(
```
Comment (shachlanAmazon) [inline redis/tests/test_cluster.rs:903]: same

File: redis/tests/test_cluster_async.rs [hunk 23]
```diff
@@ -1978,6 +1978,42 @@ fn test_async_cluster_periodic_checks_update_topology_after_failover() {
...
+        let mut connection = cluster.async_connection().await;
+        let client_info: String = cmd("CLIENT")
+            .arg("INFO")
+            .query_async(&mut connection)
+            .await
+            .unwrap();
+
+        let client_attrs = parse_client_info(&client_info);
+
+        assert!(
```
Comment (shachlanAmazon) [inline redis/tests/test_cluster_async.rs:2000]: same

**Review by shachlanAmazon — APPROVED** [review]
all comments are minor. approved.

**Review by barshaul — APPROVED** [review]
Overall approved, 
But I do think the async cluster test should route to all nodes. If it's not a lot of work adding that - please do 🕺🕺
